### PR TITLE
fix: array jmespath fail in idempotency module

### PR DESCRIPTION
**Pull request by jeromevdl**
**Issue #, if available:** #1419

## Description of changes:

Changed the implementation of `BasePersistenceStore#isMissingIdemPotencyKey()` to iterate over elements instead of fields.

**Checklist**

 

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [x] Update tests
* [ ] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/persistence/BasePersistenceStore.java [hunk 1]
```diff
@@ -223,10 +222,10 @@ public DataRecord getRecord(JsonNode data, Instant now)
...
 
-        DataRecord record = getRecord(idemPotencyKey);
-        saveToCache(record);
-        validatePayload(data, record);
-        return record;
+        DataRecord dataRecord = getRecord(idemPotencyKey);
+        saveToCache(dataRecord);
+        validatePayload(data, dataRecord);
+        return dataRecord;
     }
```
Comment (scottgerring) [inline powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/persistence/BasePersistenceStore.java:229]: Note to self: Up to this point is just refactoring variable name

File: powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/persistence/BasePersistenceStore.java [hunk 2]
```diff
@@ -258,10 +257,10 @@ private Optional<String> getHashedIdempotencyKey(JsonNode data) {
 
     private boolean isMissingIdemPotencyKey(JsonNode data) {
         if (data.isContainerNode()) {
-            Stream<Map.Entry<String, JsonNode>> stream =
-                    StreamSupport.stream(Spliterators.spliteratorUnknownSize(data.fields(), Spliterator.ORDERED),
+            Stream<JsonNode> stream =
+                    StreamSupport.stream(Spliterators.spliteratorUnknownSize(data.elements(), Spliterator.ORDERED),
```
Comment (scottgerring) [inline powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/persistence/BasePersistenceStore.java:261]: Note to self: Here's the change

**Review by scottgerring — APPROVED** [review]
LGTM
Question before we merge - 1/ how did you discover this and 2/ is there anything else in the JMESPath stuff that might be suspicious that jumped out during this change?

**Comment by jeromevdl** [comment]
> LGTM Question before we merge - 1/ how did you discover this and 2/ is there anything else in the JMESPath stuff that might be suspicious that jumped out during this change?

@scottgerring, I found this with the workshop we're doing for re:Invent. There are lots of tests but looks like we still don't cover all the branches...

**Comment by dreamorosi** [comment]
Has this been released? We have a TODO item in the workshop about this.

**Comment by jeromevdl** [comment]
Not released yet. In the next one, should be around mid-november.

**Pull request merged**